Docker Image Metadata proposal

mfojtik opened the pull request.

mfojtik commented:
@pmorie @smarterclayton @bparees 

This is **very** rough version but all comments appreciated ;-)

mfojtik commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,153 @@
+
+## The list of proposed metadata
+
+### ACCEPT_TAR_STREAM
```
Per discussion with @smarterclayton this will likely be changed to TAR_ON_INPUT or SOURCE_ON_INPUT and there will be a Trello card to support this.

bparees replied:
trello card here: https://trello.com/c/rslmL9CM/498-3-sti-support-for-images-that-accept-tar-build-future

mfojtik replied:
@smarterclayton @bparees so I think this might not be needed here after all. Since there are no images that I know if that support streaming application source as a tar to entrypoint, we can omit this metadata.

We were talking today about detecting ONBUILD instructions, which is already available in docker inspect output, so no special flag needed for that as well.

ncdc commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,153 @@
+## Motivation
+
+Image metadata are needed to do automatic generation of the OpenShift v3
+resource templates, like a PodTemplates, ReplicationControllers, BuildConfigs
```
s/like a/like/

ncdc commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,153 @@
+
+Image metadata are needed to do automatic generation of the OpenShift v3
+resource templates, like a PodTemplates, ReplicationControllers, BuildConfigs
+and other.  Automatic generation of PodTemplate is also required for
```
s/and other/, etc/

ncdc replied:
s/generation of PodTemplate/generation of a PodTemplate/

ncdc commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,153 @@
+
+## Constraints and Assumptions
+
+For the time being, there is no good way how to define metadata for Docker
```
s/how to define/to define/

ncdc commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,153 @@
+## Constraints and Assumptions
+
+For the time being, there is no good way how to define metadata for Docker
+images other than specifying extra environment variables. There is an upstream
```
s/an upstream/upstream/

ncdc commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,153 @@
+
+For the time being, there is no good way how to define metadata for Docker
+images other than specifying extra environment variables. There is an upstream
+work currently being done to fix this situlation and we should switch to that in
```
s/situlation/situation/

ncdc commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,153 @@
+
+We also assume that there might be more use-cases and metadata define for images
+that helps with generation stuff, this document only provides those needed in
+the short-term.
```
I would change this paragraph to "This document only defines the metadata needed by the current set of use cases. Additional metadata and/or use cases may be added in the future"

ncdc commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,153 @@
+## Use cases
+
+1. As an author of the Docker image, I want to specify a flag that will tell
+   whether the image I build can be used as an STI image builder, so that when
```
I assume you mean "STI builder image"?

ncdc commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,153 @@
+
+1. As an author of the Docker image, I want to specify a flag that will tell
+   whether the image I build can be used as an STI image builder, so that when
+   somebody use my image in OpenShift v3 the BuildConfig generated for this image
```
somebody use -> someone uses

ncdc replied:
add comma after v3

mfojtik commented:
@ncdc rocks! thanks, I fixed them all.

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+   somebody uses my image in OpenShift v3, the BuildConfig generated for this image
+   will default the build strategy to STI. For this, I only need to specify if my
+   entrypoint accepts a 'tar' stream that contains the application as an input.
+
```
application source?

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+
+### PUBLISHES
+
+This variable contains a list of services that the Docker image publishes. The
```
comma separated list?  semicolon?

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+
+If this variable is set, then the BuildConfig generated for this Docker image
+will default the build strategy to 'STI'. The Docker image entrypoint is
+expected to support the 'tar' input stream with the application.
```
"input stream on stdio"?

bparees replied:
Also this means we expect the image to do what needs to be done with that input such that we can then commit the new image and have a runnable application image.  So it's more than just "accepting" tar, it's going to do something intelligent.

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+'PROTO' part is optional and default to 'tcp'.
+
+When the platform consumes this Docker image, it should automatically setup
+Service(s) with given name and port, the Route should be created for this
```
does Route mean external route here? do we really always want to setup an external route?  Seems like that should be an explicit choice.

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+When the platform consumes this Docker image, it should automatically setup
+Service(s) with given name and port, the Route should be created for this
+service and the health-check should be generated for the ReplicationController
+that implements this Service.
```
"generated for the ReplicationController that manages the Pods that make up this Service"?

ryanj commented:
Will this metadata file (or a future implementation of it), be stored in a standard location inside my project repo?  

As a solution packager, I'd want to declare external dependencies using a standard filename, location, and schema.  I'd want to bundle this info (or a subset of it) inside my repo, to use as the basis for a template file.

bparees commented:
@ryanj this is image metadata, not project metadata, and it's not a file, just ENV variables defined on the image.

bparees commented on docs/proposals/metadata.md:
````diff
@@ -0,0 +1,152 @@
+distinguish between them, the proposed format for the value us following:
+
+```
+SUBSCRIBES="service:redis/6379,image:openshift/redis"
````
it bothers me that the service format here (redis/6379) does not match the publish format (redis/tcp/6379).  I think we should make them match.

mfojtik replied:
I was thinking to make 'tcp' default (less writing)

bparees replied:
that's fine, but you should do that in the publish format too then.  the syntax and behavior of both env variables should be the same, i think.

mfojtik replied:
@bparees yes, will update publishes

mfojtik replied:
@bparees actually it is:

 This variable contains a comma separated list of services that the Docker image
 publishes. The service is described as `<name>/PROTO/PORT` (eg.
 `mysql/tcp/3128`) where the 'PROTO' part is optional and default to 'tcp'.

bparees replied:
having the protocol be optional and in the middle of the stanza is going to make parsing harder/typos easier, might be better to put the optional protocol at the end (mysql/3128/tcp).

that way someone can't accidentally supply a value like "tcp/3128" which we'd consider valid

smarterclayton replied:
I don't like subscribes as a term.  It's really a soft dependency on another component.

mfojtik replied:
@smarterclayton how about SUGGESTS ?

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+publishes `redis/6379` service. If there is no match, we continue to search for
+the `openshift/redis` image.
+
+The generation tools should automatically generate corresponding Service for
```
isn't this about publish, not subscribe?  I would expect publish to trigger the generation of a Service definition, not subscribe.

bparees replied:
To clarify a bit, it seems like the flow would be:
1) pick an image that has some Subscribe definition
2) find an ImageRepository definition that fulfills the subscription
3) create a new Pod for the ImageRepo found in #2 
4) create a new Service for the service published by the Pod in #3 
5) create a new Pod for the Image in #1 
6) ??? - ensure service env variables are supplied to the pod in #5.  Today that happens automatically since they're all in the same namespace, i believe.

This also presumably assumes you never share a service (since you're always creating a new one to fulfull the subscription), so that qualification should be made clear

mfojtik replied:
@bparees subscribe (if the image is found) should automatically instantiate the 'publisher' image. Which means the service gets created, etc...

mfojtik replied:
i would say that you can't share the service... actually that is a very good point... should the generation tool be aware that there is already a service running in your namespace that fulfils the subscribe requirement?

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+
+In this case, we lookup user ImageRepositories for the Docker image that
+publishes `redis/6379` service. If there is no match, we continue to search for
+the `openshift/redis` image.
```
What happens when a match is found?

What happens when a match is not found?

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+the Docker image requires another Docker image in order to work.  For example, a
+'phpmyadmin' requires 'mysql' service to be available, otherwise it won't work
+properly. A special 'backup' Docker image for mysql will require 'mysql' to be
+available within a Pod in order to do database dump from it.
```
If an image specifies REQUIRES, can it skip SUBSCRIBES?  or must everything in REQUIRES also appear in SUBSCRIBES?  ie is the behavior of REQUIRES a superset of the behavior of SUBSCRIBES?  (that would be my preference).

mfojtik replied:
REQUIRES is 'required', iow. fail when no match found where SUBSCRIBES is 'optional', we can print suggestion but we don't break the process.

bparees replied:
how do you feel about instead extending the SUSCRIBES syntax to allow indication of "required" vs "optional", eg:  SUBSCRIBES=requires:redis/789,optional:mysql/3306

Otherwise "REQUIRES" is a bit naked...  ie it's not obvious what it refers to requiring.

If you don't like that suggestion (And i'm not saying i love it), i might also suggest:
OPTIONAL_SERVICES
REQUIRED_SERVICES
as the subscription env variables instead.

or OPTIONAL_SUBSCRIPTIONS, REQUIRED_SUBSCRIPTIONS

smarterclayton replied:
So this shouldn't be tied to services, this should be tied to opaque functionality.  It really has nothing to do with services.  At a bare minimum, it has to be tied to a string that represents something understandable to a user.

I don't think we should do SUBSCRIBES today, I don't think that's required in scope for 3.0

bparees replied:
what else provides "opaque functionality" other than services, and how can we support that in a generic, yet programmatically useful, way?  I think the existing proposal solves a valid common use case..

smarterclayton replied:
We already tried this with v2, there's a long history, but basically "explicit name lookup" fails.  If your image ootb needs other images, create a template.  If you need us to figure out other things based on name, we actually need to go figure out what "mysql" means to you, because that's not enough info to do the right thing (every time).  I don't think the "i need a database" is a use case for an image right now - instead the more valuable thing is "which of these images is a mysql database" and "how does the UI go figure out which things are mysql databases."  That's the metadata that is more valuable right now.  

----- Original Message -----
> > +In this case, we lookup user ImageRepositories for the Docker image that
> > +publishes `redis/6379` service. If there is no match, we continue to
> > search for
> > +the `openshift/redis` image.
> > +
> > +The generation tools should automatically generate corresponding Service
> > for
> > +each defined service.
> > +
> > +### REQUIRES
> > +
> > +This variable contains a list of services that the Docker image requires
> > in
> > +order to work properly. The value format is same as for SUBSCRIBES.  Where
> > the
> > +SUBSCRIBES setup the infrastructure for pub/sub the REQUIRES might
> > indicate that
> > +the Docker image requires another Docker image in order to work.  For
> > example, a
> > +'phpmyadmin' requires 'mysql' service to be available, otherwise it won't
> > work
> > +properly. A special 'backup' Docker image for mysql will require 'mysql'
> > to be
> > +available within a Pod in order to do database dump from it.
> 
> what else provides "opaque functionality" other than services, and how can we
> support that in a generic, yet programmatically useful, way?  I think the
> existing proposal solves a valid common use case..
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24195096

bparees replied:
"which of these images is a mysql DB?"  the one that PUBLISHES a mysql service.

smarterclayton replied:
What about 5.1.  What about percona?  What about a mysql slave?  We've gone down this road before, there's no magic solution.

----- Original Message -----
> > +In this case, we lookup user ImageRepositories for the Docker image that
> > +publishes `redis/6379` service. If there is no match, we continue to
> > search for
> > +the `openshift/redis` image.
> > +
> > +The generation tools should automatically generate corresponding Service
> > for
> > +each defined service.
> > +
> > +### REQUIRES
> > +
> > +This variable contains a list of services that the Docker image requires
> > in
> > +order to work properly. The value format is same as for SUBSCRIBES.  Where
> > the
> > +SUBSCRIBES setup the infrastructure for pub/sub the REQUIRES might
> > indicate that
> > +the Docker image requires another Docker image in order to work.  For
> > example, a
> > +'phpmyadmin' requires 'mysql' service to be available, otherwise it won't
> > work
> > +properly. A special 'backup' Docker image for mysql will require 'mysql'
> > to be
> > +available within a Pod in order to do database dump from it.
> 
> "which of these images is a mysql DB?"  the one that PUBLISHES a mysql
> service.
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24210849

bparees replied:
i'm not clear how you're planning to solve the problem you posed, though, "which of these images is a mysql database", if you remove this metadata.  It's still a name-matching problem.

smarterclayton replied:
Difference between matching and equality.

----- Original Message -----
> > +In this case, we lookup user ImageRepositories for the Docker image that
> > +publishes `redis/6379` service. If there is no match, we continue to
> > search for
> > +the `openshift/redis` image.
> > +
> > +The generation tools should automatically generate corresponding Service
> > for
> > +each defined service.
> > +
> > +### REQUIRES
> > +
> > +This variable contains a list of services that the Docker image requires
> > in
> > +order to work properly. The value format is same as for SUBSCRIBES.  Where
> > the
> > +SUBSCRIBES setup the infrastructure for pub/sub the REQUIRES might
> > indicate that
> > +the Docker image requires another Docker image in order to work.  For
> > example, a
> > +'phpmyadmin' requires 'mysql' service to be available, otherwise it won't
> > work
> > +properly. A special 'backup' Docker image for mysql will require 'mysql'
> > to be
> > +available within a Pod in order to do database dump from it.
> 
> i'm not clear how you're planning to solve the problem you posed, though,
> "which of these images is a mysql database", if you remove this metadata.
> It's still a name-matching problem.
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24211200

mfojtik replied:
@smarterclayton if there will be no metadata that would say "this image is exposing mysql service kind of service", how you gonna know what image to choose other than asking user what image he wants?

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+
+This variable contains a list of services that are companions to the Docker
+image. For example a 'redis-slave' Docker image is a companion for
+'redis-master' Docker image. Where the SUBSCRIBES defines relations to services
```
wait, if "subscribes" lists services that are required, what's the point of "requires"?

mfojtik replied:
COMPANIONS are going away, they are the same as SUBSCRIBES.

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+
+### SECRETS
+
+This variable contains a list of secrets (secret indentifiers) that the Docker
```
s/indentifiers/identifiers/

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+generation tools should check if the required secrets are available and if not,
+it should ask the user to provide them.
+
+An example value can be: `SECRETS=keys/twitter,keys/facebook`
```
Need more examples here of how different types of secrets are handled (eg password value vs a certificate file), and how secrets are provided to the image.

Maybe this doc isn't the place for it, but it definitely needs elaboration.

mfojtik replied:
@bparees @pmorie there is a PR opened for kubernetes with secrets proposal. Secrets will be most likely mounted as volume.

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+This variable indicates if the Docker image is clusterable or not. In case of
+Docker image with standalone 'mysql' server you don't want to have the number of
+replicas in ReplicationController be higher than 1, otherwise you can risk that
+multiple standalone mysql servers will access data on shared volume.
```
actually the multiple mysql servers would each have their own volume, but that is also problematic if the assumption is that the service represents a single DB.

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+
+This variable contains a list of Docker volumes that will be host-mounted
+instead of network mounted. These volumes will not be persistent, but they might
+be faster than network volumes.
```
so is it assumed that the existing docker VOLUMES metadata will be mapped as networked/shared volumes?  if so, let's state that.

mfojtik replied:
yeah, good catch.

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+This variable contains a list of URL mappings. For example, if your 'phpmyadmin'
+is available on '/phpmyadmin' URL, you can override this by specifying a mapping
+in following form: `/phpmyadmin:/`. In this case the traffic will gets routed
+from '/phpmyadmin' to '/'.
```
this example is confusing.  "if your phpmyadmin is available on /phpmyadmin url" - available where?  I think this is trying to say "if the image services requests on http://container.ip/phpmyadmin, then /phpmyadmin:/ will result in external requests to "http://router.ip/" being routed to http://container.ip/phpmyadmin ", yes?

If so, in addition to clarifying the description, I think the mapping order is backwards.  We are mapping / to /phpmyadmin, so the mapping should be "/:/phpmyadmin"

On the otherhand if this is really trying to express that the image should be exposed externally as http://router.ip/phpmyadmin but that the container itself handles requests at http://container.ip/, then the mapping order makes sense.  Not clear which was intended.

Either way, the intent is that this mapping feeds into the generation of a Router definition, right? so we should state that.  (And what's the implication for Services?  Services don't allow url mapping as far as i know)

mfojtik replied:
it makes me think if the URL_MAPPINGS are necessary... I mean you as an image author can define on what URI you will server your app (and you can also setup rewrite/redirects).

bparees replied:
hm, well you still might want to setup mappings at the Router layer, right?  So being able to define those mappings via image metadata makes sense to me so we can auto-generate the right Router definition.

smarterclayton replied:
I don't think we need this for 3.0

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+in following form: `/phpmyadmin:/`. In this case the traffic will gets routed
+from '/phpmyadmin' to '/'.
+
+### MIN_CPU,MIN_MEMORY,MIN_###
```
Should that last item be MIN_XXX if you're trying to imply "and other categories?

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+
+### MIN_CPU,MIN_MEMORY,MIN_###
+
+This variable define a minimal amount of resources needed in order to make this
```
s/define/defines/

bparees commented:
@mfojtik comments made.  will await appreciation.

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+This variable contains a list of Docker volumes that will be host-mounted
+instead of network mounted. These volumes will not be persistent, but they might
+be faster than network volumes. If the volumes are not specified, then they are
+mounted as network volumes.
```
I don't think we need this.

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+This variable indicates if the Docker image is clusterable or not. In case of
+Docker image with standalone 'mysql' server you don't want to have the number of
+replicas in ReplicationController be higher than 1 if the assumption is that the
+Service that mysql provides represents a single database.
```
This is a suggestion, not a hard requirement.  The UIs will only suggest that you don't cluster this, they won't prevent you.

mfojtik replied:
@smarterclayton is there a way to set the hard limit of maximum number of replicas?

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+will default the build strategy to 'STI'. The Docker image entrypoint is
+expected to support the 'tar' input stream with the application.
+
+### PUBLISHES
```
This has nothing to do with services in Kubernetes.  I don't think we should do anything other than decorate the ports with metadata, but it should be port specific.

bparees replied:
how does this not have to do w/ services in k8s?  the intent of this is to let us programmatically generate Service definitions appropriate to an Image that's being deployed in a Pod.

smarterclayton replied:
We already have that, it's "use the first port in the specs".

----- Original Message -----
> > +   Image.
> > +
> > +5. As an author of the Docker image, I need to have a way to disable
> > +   'clustering' for my Image, in case the Docker image does not support
> > this.
> > +   In other words, if the Docker image specify this, the
> > ReplicationController
> > +   generated for this Image should never get `replicas` set higher than 1.
> > +
> > +## The list of proposed metadata
> > +
> > +### ACCEPT_TAR_STREAM
> > +
> > +If this variable is set, then the BuildConfig generated for this Docker
> > image
> > +will default the build strategy to 'STI'. The Docker image entrypoint is
> > +expected to support the 'tar' input stream with the application.
> > +
> > +### PUBLISHES
> 
> how does this not have to do w/ services in k8s?  the intent of this is to
> let us programmatically generate Service definitions appropriate to an Image
> that's being deployed in a Pod.
> 
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24195419

mfojtik replied:
how do you know what is the service name? how do you know what kind of service it will be?

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+publishes. The service is described as `<name>/PROTO/PORT` (eg.
+`mysql/tcp/3128`) where the 'PROTO' part is optional and defaults to 'tcp'.
+
+When the platform consumes this Docker image, it should automatically setup
```
I don't think this is name based - the generation flows should simply suggest to the user a port that will be exposed, and offer options.  We don't need proscriptive settings in the metadata.

Simplify this down to port metadata.

bparees replied:
that seems like it's going to lead to a lot of mandatory prompting/wizards that we could avoid with better metadata.

smarterclayton replied:
It doesn't have to prompt, it just has to make a reasonable guess, then let you tweak it.  Ports are something that change all the time, and we never get them right.  So I'm less concerned about the magic here - the first exposed port is what we would give you, then if that doesn't work.... you go and change the port.

----- Original Message -----
> > +
> > +## The list of proposed metadata
> > +
> > +### ACCEPT_TAR_STREAM
> > +
> > +If this variable is set, then the BuildConfig generated for this Docker
> > image
> > +will default the build strategy to 'STI'. The Docker image entrypoint is
> > +expected to support the 'tar' input stream with the application.
> > +
> > +### PUBLISHES
> > +
> > +This variable contains a comma separated list of services that the Docker
> > image
> > +publishes. The service is described as `<name>/PROTO/PORT` (eg.
> > +`mysql/tcp/3128`) where the 'PROTO' part is optional and defaults to
> > 'tcp'.
> > +
> > +When the platform consumes this Docker image, it should automatically
> > setup
> 
> that seems like it's going to lead to a lot of mandatory prompting/wizards
> that we could avoid with better metadata.
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24195541

bparees replied:
i really think we can do better than that.  the ecosystem of openshift images will be much stronger if we enable people to add this sort of metadata to them.  how many developers are going to understand how to consume an image if they have to go figure out what port/protocol they need to map a service to?

smarterclayton replied:
We pick the first one and tell them if there are others, then you make sure that it's predictable each time.  Your protocol information helps, but you can't solve this generically.

----- Original Message -----
> > +
> > +## The list of proposed metadata
> > +
> > +### ACCEPT_TAR_STREAM
> > +
> > +If this variable is set, then the BuildConfig generated for this Docker
> > image
> > +will default the build strategy to 'STI'. The Docker image entrypoint is
> > +expected to support the 'tar' input stream with the application.
> > +
> > +### PUBLISHES
> > +
> > +This variable contains a comma separated list of services that the Docker
> > image
> > +publishes. The service is described as `<name>/PROTO/PORT` (eg.
> > +`mysql/tcp/3128`) where the 'PROTO' part is optional and defaults to
> > 'tcp'.
> > +
> > +When the platform consumes this Docker image, it should automatically
> > setup
> 
> i really think we can do better than that.  the ecosystem of openshift images
> will be much stronger if we enable people to add this sort of metadata to
> them.  how many developers are going to understand how to consume an image
> if they have to go figure out what port/protocol they need to map a service
> to?
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24210697

bparees replied:
we can at least solve it for the images we provide out of the box which will provide a good sample for future image authors.

mfojtik replied:
+1 @bparees

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+   the image is deployed, we automatically setup a Service for it, add health-check
+   for 'mysql' service type and wire the Route for the 3128 port.
+
+3. As an user of OpenShift v3 I want to be able to say what services my
```
I don't think this is a 3.0 use case.

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+   my application, so the platform will check the ImageRepositories for the
+   Image that have `PUBLISHES="mysql/tcp/3128"` and add it to my deployment.
+
+4. As an author of the Docker image, I want to specify what _secrets_ the image
```
I don't think this is a 3.0 requirement, and I'm unclear that this should even be supported.  Can this be handled better by documentation on image startup if secrets aren't available?

bparees replied:
this is going to be the mechanism for getting DB passwords and such, is it not?

smarterclayton replied:
That mechanism is templates.

----- Original Message -----
> > +
> > +2. As an author of the Docker image, I want to specify list of services
> > that the
> > +   image I'm building will provide, so in OpenShift v3 we can
> > automatically
> > +   setup Routes, Services and health-checks for this image. For example,
> > in case
> > +   I'm building 'mysql' image, I can specify `PUBLISHES=mysql/tcp/3128` so
> > when
> > +   the image is deployed, we automatically setup a Service for it, add
> > health-check
> > +   for 'mysql' service type and wire the Route for the 3128 port.
> > +
> > +3. As an user of OpenShift v3 I want to be able to say what services my
> > +   application will consume, so the platform can find the corresponding
> > Docker
> > +   images and make the services available for my application.
> > +   For example, as an user I can say I want to have 'mysql' service
> > available in
> > +   my application, so the platform will check the ImageRepositories for
> > the
> > +   Image that have `PUBLISHES="mysql/tcp/3128"` and add it to my
> > deployment.
> > +
> > +4. As an author of the Docker image, I want to specify what _secrets_ the
> > image
> 
> this is going to be the mechanism for getting DB passwords and such, is it
> not?
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24196840

bparees replied:
so i have to write a template anytime i want to share a password between two images?

smarterclayton replied:
For 3.0, sure.

----- Original Message -----
> > +
> > +2. As an author of the Docker image, I want to specify list of services
> > that the
> > +   image I'm building will provide, so in OpenShift v3 we can
> > automatically
> > +   setup Routes, Services and health-checks for this image. For example,
> > in case
> > +   I'm building 'mysql' image, I can specify `PUBLISHES=mysql/tcp/3128` so
> > when
> > +   the image is deployed, we automatically setup a Service for it, add
> > health-check
> > +   for 'mysql' service type and wire the Route for the 3128 port.
> > +
> > +3. As an user of OpenShift v3 I want to be able to say what services my
> > +   application will consume, so the platform can find the corresponding
> > Docker
> > +   images and make the services available for my application.
> > +   For example, as an user I can say I want to have 'mysql' service
> > available in
> > +   my application, so the platform will check the ImageRepositories for
> > the
> > +   Image that have `PUBLISHES="mysql/tcp/3128"` and add it to my
> > deployment.
> > +
> > +4. As an author of the Docker image, I want to specify what _secrets_ the
> > image
> 
> so i have to write a template anytime i want to share a password between two
> images?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24210577

mfojtik replied:
@smarterclayton isn't this about generating templates? so when then image has SECRETS, we add secrets volume to the template? We can also check if the user has the secrets available, if not we can ask him to add them? (twitter api?)

mfojtik replied:
@smarterclayton metadata is "I as an image author need these secrets mounted to make this image useful", template is more "based on the metadata, I'm going to generate PodTemplate and VOLUMES that satisfy this"

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+5. As an author of the Docker image, I need to have a way to disable
+   'clustering' for my Image, in case the Docker image does not support this.
+   In other words, if the Docker image specify this, the ReplicationController
+   generated for this Image should never get `replicas` set higher than 1.
```
An image may suggest that it does not support scaling, and the UI will communicate that to end consumers.  It is not a requirement and should not be enforced at the UI.

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,152 @@
+in following form: `/phpmyadmin:/`. In this case the traffic will gets routed
+from '/phpmyadmin' to '/'.
+
+### MIN_CPU,MIN_MEMORY,MIN_RESOURCE
```
Suggestions only.

mfojtik commented:
@smarterclayton @bparees so after all, it only required metadata we need for 3.0 is the 'PUBLISHES'. Right now the PUBLISHES takes a list of services that maps to the ports: `mysql/3128/tcp`. This basically adds a context to the EXPOSE 3128/tcp that this port provides the 'mysql' service. 

I think the last open question was how we are going to name this metadata? I like PUBLISHES, because in future it might match with SUBSCRIBES. It is also generic, so I don't see any problems with upstream adopting this name.

bparees commented:
Clayton is still not on board w/ the publishes value/format and i think he's semi-convinced me (or worn me down anyway).  so i'm not sure what is left at all :)

bparees commented:
I think @smarterclayton would rather see a more generic "WANTS=<generic string/tag>" and "PROVIDES=<generic string/tag> " where values might be "database" or "mysql"

smarterclayton commented:
We should make a decision from the UI side whether we want to have two sets of "things that are relevant".  In OpenShift today we have categories (tags) and provides (structured requirements).  The UI uses the tags heavily, and while provides are used under the covers, they are not exposed to clients.  I would argue for reserving PROVIDES for "structured requirements" and not implement it now, and instead do CATEGORIES / TAGS.  I don't think we have a hard requirement on WANTS (loose) without structured requirements - instead the UI merely suggests things to you based on what you have (TAGS you have already vs TAGS you don't have).  Punting on structured requirements simplifies the problem, and it's more of a client side organizational concept.

----- Original Message -----
> I think @smarterclayton would rather see a more generic "WANTS=<generic
> string/tag>" and "PROVIDES=<generic string/tag> " where values might be
> "database" or "mysql"
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906#issuecomment-73557217

mfojtik commented:
@smarterclayton @bparees stripped the proposal to minimum :-) Marked items that are post 3.0, please let me know if you want to see more use cases or add more informations into the proposed metadata variables.

bparees commented:
hmm, do we not even need to know if a port is tcp vs udp?  are we going to support udp proxying/routing?

soltysh commented:
My only worry here is that we removed information about STI compatibility, which IIRC was one of the main points. I'm aware that with recent @mfojtik work on STI (https://github.com/openshift/source-to-image/pull/124) we'll be able to consume more images as STI builders, still I think we should promote STI compatibility.

mfojtik commented:
@soltysh

bparees commented:
@soltysh STI still defines the two STI env variables (scripts url and location), if they're there we'll use them in our generation logic,  so yeah they might be worth mentioning here, but they aren't something openshift is really defining.

mfojtik commented:
@bparees +1 they are something that we can recognize in our generation tools already, no need to 'define' them.

mfojtik commented:
@bparees the EXPOSE has syntax for that:

```
EXPOSE 1234/tcp
```

bparees commented:
@mfojtik ok, i did a quick glance at the dockerfile docs and didn't see that syntax.  sounds good.

smarterclayton commented:
Services already support udp

smarterclayton commented:
Agreed


> On Feb 11, 2015, at 7:00 AM, Maciej Szulik <notifications@github.com> wrote:
> 
> My only worry here is that we removed information about STI compatibility, which IIRC was one of the main points. I'm aware that with recent @mfojtik work on STI (openshift/source-to-image#124) we'll be able to consume more images as STI builders, still I think we should promote STI compatibility.
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,80 @@
+### TAGS
+
+This metadata variable contains a list of tags which maps to the ports that the
+Docker image expose (EXPOSE instruction). The TAGS provides the context to the
```
Tags doesn't have anything to do with ports really.  Just say that tags are a way to categorize images into broad areas of functionality

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,80 @@
+Being not-scalable basically means that the value of 'replicas' should initially
+not be set higher than 1.
+
+### MIN_CPU,MIN_MEMORY,MIN_RESOURCE
```
Given the sophistication of the Kube resource scheme, we may find ourselves in a trap here.  We should define some example values like "compatible with the Kube resource quantity and values for kcpu and memory".  Reference the exact Kube resource amounts.  Give an example.

Nuke min_resource

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,80 @@
+
+## Metadata variables (required for 3.0)
+
+### TAGS
```
We need a common prefix that won't collide with app env.  Something like "_IMAGE_"

mfojtik replied:
@smarterclayton I like IMAGE_

smarterclayton replied:
The prefix is probably going to have to be more unique - if we start colliding with values end users provide that's bad.

----- Original Message -----
> > +   platform, I want to express what categories/tags my Docker image will
> > belongs
> > +   to, so the platform then can then consume the tags to improve the
> > generation
> > +   workflow based on them.
> > +
> > +2. As an user of OpenShift I want to get reliable suggestions about
> > services
> > +   that the Docker image I'm going to use might require to operate
> > properly. As
> > +   an author of the Docker image, I need to have a way to record what
> > services my
> > +   Docker image want to consume.
> > +
> > +2. As an author of the Docker image, I need to have a way to indicate
> > whether
> > +   the container started from my Docker image does not support scaling.
> > +   The UI should then communicate this information with the end consumers.
> > +
> > +## Metadata variables (required for 3.0)
> > +
> > +### TAGS
> 
> @smarterclayton I like IMAGE_
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24571508

smarterclayton commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,80 @@
+### MIN_CPU,MIN_MEMORY,MIN_RESOURCE
+
+This variable suggests how much resources the Docker image might need in order
+to work properly. The UI might warn the user that deploying this Docker image
```
Where did multiple port protocols go?

mfojtik replied:
I thought you want to nuke PUBLISHES?

smarterclayton replied:
I don't want the linkage aspect of publishes, I want the port metadata.  It's additions to EXPOSE.  I would call it out as something that is very specifically as additional expose metadata.

----- Original Message -----
> > +
> > +The format of the values is the same as for TAGS.
> > +
> > +## Metadata variables (post 3.0)
> > +
> > +### SINGLETON
> > +
> > +An image might use this variable to suggest that it does not support
> > scaling.
> > +The UI will then communicate this to consumers of that image.
> > +Being not-scalable basically means that the value of 'replicas' should
> > initially
> > +not be set higher than 1.
> > +
> > +### MIN_CPU,MIN_MEMORY,MIN_RESOURCE
> > +
> > +This variable suggests how much resources the Docker image might need in
> > order
> > +to work properly. The UI might warn the user that deploying this Docker
> > image
> 
> I thought you want to nuke PUBLISHES?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24571631

bparees commented:
@smarterclayton just mentioned we should have a DESCRIPTION field.

mfojtik commented:
@bparees @smarterclayton why not ANNOTATIONS ?

bparees commented:
@mfojtik instead of description?

ncdc commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,90 @@
+
+An image might use this variable to suggest that it does not support scaling.
+The UI will then communicate this to consumers of that image.
+Being not-scalable basically means that the value of 'replicas' should initially
```
initially or ever?

mfojtik replied:
@ncdc there is nothing that prevents you from scaling up, right? this should be just a suggestion to UI.

ncdc replied:
True, I suppose there's not a way to prevent it.

smarterclayton replied:
Today in OpenShift v2 it's a suggestion on quick starts.

----- Original Message -----
> > +
> > +Specifies a list of tags that the generation tools and UI might use for
> > +suggesting to provide if you don't have the Docker images with given tags
> > +already. For example, if the Docker image WANTS 'mysql' and 'redis' and
> > you
> > +don't have the Docker image with 'redis' tag, then UI might suggest you to
> > add
> > +this image into your application.
> > +
> > +The format of the values is the same as for TAGS.
> > +
> > +## Metadata variables (post 3.0)
> > +
> > +### IMAGE_NON_SCALABLE
> > +
> > +An image might use this variable to suggest that it does not support
> > scaling.
> > +The UI will then communicate this to consumers of that image.
> > +Being not-scalable basically means that the value of 'replicas' should
> > initially
> 
> True, I suppose there's not a way to prevent it.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24818898

smarterclayton commented:
It's a pretty technical term - may not be recognizable to someone glancing at it.

----- Original Message -----
> @bparees @smarterclayton why not ANNOTATIONS ?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906#issuecomment-74660113

mfojtik commented:
@smarterclayton so it will be used to describe the image functionality? Like 

DESCRIPTION="MySQL 5.5 image with advanced logging"

mfojtik commented:
@smarterclayton @bparees added IMAGE_DESCRIPTION and IMAGE_EXPOSE_SERVICES 

I think we collected all metadata that we require for 3.0, so can you please do a final review? Once this is pushed, we should start adding the metadata into STI/database images we are working on right now.

bparees commented:
@mfojtik lgtm.

smarterclayton commented on docs/proposals/metadata.md:
````diff
@@ -0,0 +1,118 @@
+Example:
+
+```
+IMAGE_EXPOSE_SERVICES="3128/tcp:mysql,8080:http"
````
I don't think this is right.  I would prefer IMAGE_EXPOSE_<port>_PROTOCOLS="protocol1,protocol2,...".

smarterclayton commented:
If image metadata makes docker 1.6 we probably will use those directly, and so will instead want to determine how each of these will map to standard image metadata keys (do we need a namespace, what are the Kube values, etc).

mfojtik commented:
@smarterclayton do you have link to Docker PR with image metadata?

smarterclayton commented:
https://github.com/GoogleCloudPlatform/kubernetes/issues/3764

----- Original Message -----
> @smarterclayton do you have link to Docker PR with image metadata?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906#issuecomment-86498020

mfojtik commented:
@smarterclayton the LABEL is available in docker-1.6 which was released recently (I've got it on my F22 today, but not everyone is updated now). Should we wait till this update lands, or I can start switching our images to use LABEL(s) instead of ENV now?

smarterclayton commented:
Once it hits for centos let's start switching.  Coordinate with Andy.  Also, we need to sync this doc with the work that the appinfra team is doing around the atomic command and the LABELS they have defined.


> On Apr 8, 2015, at 3:03 PM, Michal Fojtik <notifications@github.com> wrote:
> 
> @smarterclayton the LABEL is available in docker-1.6 which was released recently (I've got it on my F22 today, but not everyone is updated now). Should we wait till this update lands, or I can start switching our images to use LABEL(s) instead of ENV now?
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
@smarterclayton one last refinement to the proposal... I think it is more reasonable to rename the metadata to:

`IMAGE_DESCRIPTION` => `description`
`RUBY_VERSION` => `ruby_version`
`IMAGE_TAGS` => `tags`
`IMAGE_EXPOSED_SERVICES` => `exposed_services`

What do you think? Is it necessary to have everything uppercase? It is necessary to have 'IMAGE_' prefix? I don't think that we will collide with other labels and if we do, then I think `tags` will be pretty much what we use. Same with `description`.

smarterclayton commented:
I don't have an issue with that since they will be used in labels.

However, generally we use camelCase or dashed-names for variables in visual output. What do labels allow?

> On Apr 21, 2015, at 7:06 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> @smarterclayton one last refinement to the proposal... I think it is more reasonable to rename the metadata to:
> 
> IMAGE_DESCRIPTION => description
> RUBY_VERSION => ruby_version
> IMAGE_TAGS => tags
> IMAGE_EXPOSED_SERVICES => exposed_services
> 
> What do you think? Is it necessary to have everything uppercase? It is necessary to have 'IMAGE_' prefix? I don't think that we will collide with other labels and if we do, then I think tags will be pretty much what we use. Same with description.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
@smarterclayton I think LABEL allow all syntaxes, but in the official Docker docs they use 'javish' "com.something.foo" syntax. I think dashed names is something I like here ;-)

smarterclayton commented:
The thing about picking a name format is that we have to be able to use it in annotations eventually.  So I'm a bit concerned about picking an arbitrary one without agreement from Kube.

> On Apr 21, 2015, at 8:07 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> @smarterclayton I think LABEL allow all syntaxes, but in the official Docker docs they use 'javish' "com.something.foo" syntax. I think dashed names is something I like here ;-)
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
@smarterclayton I can go ahead and open an issue on kube to gather this feedback. I think either camelCase or use dash is reasonable. (btw. I updated this PR with changes and I think we should merge it at some point, as this is hanging here for couple months ;-)

smarterclayton commented:
You should canvas all our existing usage of annotations in Openshift and identify our disrepancies.  Our annotations should typically be namespaced, although not all scenarios do.

> On Apr 21, 2015, at 8:27 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> @smarterclayton I can go ahead and open an issue on kube to gather this feedback. I think either camelCase or use dash is reasonable. (btw. I updated this PR with changes and I think we should merge it at some point, as this is hanging here for couple months ;-)
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
@smarterclayton with 'our' annotations you mean the openshift specific annotations? I don't see 'description' as something OpenShift specific. I think many Docker image will have this annotation already set and we can just profit from it.

smarterclayton commented:
Yes, all of the ones we use today (which may have inconsistent conventions). Description is Kube is probably going to be "kubernetes.io/description" or "k8s.io/description", along with other fields like displayName

> On Apr 21, 2015, at 8:40 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> @smarterclayton with 'our' annotations you mean the openshift specific annotations? I don't see 'description' as something OpenShift specific. I think many Docker image will have this annotation already set and we can just profit from it.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
@smarterclayton so even the labels will be namespaced? I was more thinking that folks which writes Docker images and push them to Docker Hub will probably not use these but rather just set the plain 'description' field.

mfojtik commented:
@smarterclayton also we are talking here about the "Docker" image metadata that we put on top of the image to make it more consumable by OpenShift and have more description about it. Does this mean that k8s is going to push Docker to use their label format (if you want to use this image with kubernetes?). Doesn't this cause a huge mess in the end? (iow. every vendor will have custom metadata names?)

smarterclayton commented:
Trying to take over the global namespace is even more worrying - it means everyone fights to get the same variables, and means we don't understand exactly how others are using those variables.

Simple things like description and display name ("display-name", "displayName", "DisplayName", or "displayname"?) might end up being standard.  The more complex fields will almost certainly not, and it's better to properly namespace them than to try to fight for limited names.

----- Original Message -----
> @smarterclayton also we are talking here about the "Docker" image metadata
> that we put on top of the image to make it more consumable by OpenShift and
> have more description about it. Does this mean that k8s is going to push
> Docker to use their label format (if you want to use this image with
> kubernetes?). Doesn't this cause a huge mess in the end? (iow. every vendor
> will have custom metadata names?)
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906#issuecomment-96607376

mfojtik commented:
@smarterclayton the Docker guidelines about labels says that the dashes should not be used:

> Keys should only consist of lower-cased alphanumeric characters, dots and dashes (for example, [a-z0-9-.])

I reworked this proposal to also include `displayName`, added notion about namespaces and polished the text to be LABEL friendly.

Can you please take a look? Also I'm waiting now how the discussion about LABEL standards ends up. However, in the mean time, do you think we can merge this and switch our Docker image to use the LABELS we proposed here? I don't want to do this as last-minute change as this might break people demos, web console, imageStream imports, etc...

bparees commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,172 @@
+[`expose-services`](#expose-services) | []string | openshift.io
+[`non-scale`](#non-scale)             |     bool | openshift.io
+[`min-cpu`](#min-cpu)                 |   string | openshift.io(?)
+[`min-memory`](#min-memory)           |   string | openshift.io(?)
```
i'd expect most of these, except maybe tags and wants, to be in the k8s.io namespace.

mfojtik replied:
@bparees that is why I put question mark on the last two (they are post-3.0 anyway)

For 'non-scale' and 'expose-services' I'm not sure. The nonScale will be primarily used in OpenShift UI to indicate that this image cannot be scaled with replication controller (so the number of replicas should be 1). As we discussed before, this will be only suggestion as nothing prevents you to scale it up....

For expose-services, yeah, that might end up as k8s.io namespace if Kubernetes is going to pick it up and use it for something. @smarterclayton do you know if there are any plans on kube side to use this metadata? (like to setup livenessprobe?)

smarterclayton replied:
> On Apr 29, 2015, at 4:35 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> In docs/proposals/metadata.md:
> 
> > +Kubernetes the namespace is `k8s.io/`. For simple labels, like `displayName` or
> > +`description` there might be no namespace set if they end up as standard in
> > +Docker.
> > +
> > +## Image Metadata
> > +
> > +Name                                  | Type     | Target Namespace |
> > +--------------------------------------|--------- | ------------------
> > +[`tags`](#tags)                       | []string | openshift.io
> > +[`wants`](#wants)                     | []string | openshift.io
> > +[`display-name`](#display-name)       |   string | k8s.io
> > +[`description`](#description)         |   string | k8s.io
> > +[`expose-services`](#expose-services) | []string | openshift.io
> > +[`non-scale`](#non-scale)             |     bool | openshift.io
> > +[`min-cpu`](#min-cpu)                 |   string | openshift.io(?)
> > +[`min-memory`](#min-memory)           |   string | openshift.io(?)
> @bparees that is why I put question mark on the last two (they are post-3.0 anyway)
> 
> For 'non-scale' and 'expose-services' I'm not sure. The nonScale will be primarily used in OpenShift UI to indicate that this image cannot be scaled with replication controller (so the number of replicas should be 1). As we discussed before, this will be only suggestion as nothing prevents you to scale it up....
> 
> For expose-services, yeah, that might end up as k8s.io namespace if Kubernetes is going to pick it up and use it for something. @smarterclayton do you know if there are any plans on kube side to use this metadata? (like to setup livenessprobe?)
> 
Maybe not as it currently is, but with some modification.  We don't have a lot of bandwidth available upstream for discussing it though until the other issues are closed.  We can always change post 3.0

> —
> Reply to this email directly or view it on GitHub.
>

mfojtik replied:
@smarterclayton so can we go ahead and start changing our Docker images now? (and maybe merge this veteran PR? ;-)

bparees replied:
what happens when a docker 1.5 client pulls an image with labels?  

also we can't remove the existing ENV "labels" until we update the openshift code to look at the docker labels instead of the ENV variables, so this probably needs to be a 3 step process.  (and we can't do it until we've moved to a 1.6 prereq)

ncdc replied:
I assume 1.5 won't see the labels.

We require 1.6 now (v2 registry PR merged). Unfortunately we're still fixing some build/release issues, but 1.6 is now a hard minimum requirement.

bparees commented:
@mfojtik one minor discussion point, otherwise i like this and can't wait to move to docker labels instead of env variables :)

soltysh commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,172 @@
+## Use cases
+
+1. As an author of Docker image that is going to be consumed by the OpenShift
+   platform, I want to express what categories/tags my Docker image will belongs
```
will belong~~s~~ to

soltysh commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,172 @@
+2. As an user of OpenShift I want to get reliable suggestions about services
+   that the Docker image I'm going to use might require to operate properly. As
+   an author of the Docker image, I need to have a way to record what services my
+   Docker image want to consume.
```
wants to consume

soltysh commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,172 @@
+
+3. As an author of the Docker image, I need to have a way to indicate whether
+   the container started from my Docker image does not support scaling.
+   The UI should then communicate this information with the end consumers.
```
communicate this information to the end users

soltysh commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,172 @@
+
+4. As an author of the Docker image, I need to have a way to indicate what
+   additional service(s) my Docker image might need to work properly, so the UI
+   or generation tools can suggest them to the end users.
```
to ~~the~~ end users

soltysh commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,172 @@
+
+### `tags`
+
+This label contains a list of tags represented as list of comma separated string
```
This label contains a list of comma separated string values.

soltysh commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,172 @@
+
+This label contains a list of tags represented as list of comma separated string
+values. The tags are the way to categorize the Docker images into broad areas of
+functionality. Tags helps UI and generation tools to suggest relevant Docker
```
Tags help~~s~~

soltysh commented on docs/proposals/metadata.md:
```diff
@@ -0,0 +1,172 @@
+
+This label suggests how much resources the Docker image might need in order
+to work properly. The UI might warn the user that deploying this Docker image
+may exceed the user quota.
```
exceed their user quota

mfojtik commented:
@soltysh @bparees @smarterclayton fixed last comments, I going to merge this now (we can update later if necessary), so I can reference to this in other docs.

[merge]
